designs/disruption-termination-grace-period.md

```diff
@@ -0,0 +1,122 @@
+# Disruption Termination Grace Period
+
+## Motivation
+Users are requesting the ability to control how long karpenter will wait for the deletion of an individual node to complete before forcefully terminating the node regardless of pod status. [#743](https://github.com/kubernetes-sigs/karpenter/issues/743). This supports two primary use cases.
+* Cluster admins who want to ensure that nodes are cycled after a given period of time, regardless of user-defined disruption controls (such as PDBs or preStop hooks) that might prevent eviction of a pod beyond the configured limit. This could be to satisfy security requirements or for convenience.
+* Cluster admins who want to allow users to protect long-running jobs from being interrupted by node disruption, up to a configured limit.
+
+This design piggybacks off the existing disruption-controls design, adding in a new `terminationGracePeriod` field to the `v1beta/NodePool` `Disruption` spec. This field seems intuitively related to the other disruption controls that exist within the disruption block.
+
+## Proposed Spec
+
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenUnderutilized || WhenEmpty
+    consolidateAfter: 10m || Never # metav1.Duration
+    expireAfter: 10m || Never # Equivalent to v1alpha5 TTLSecondsUntilExpired
+    terminationGracePeriod: 24h || nil # metav1.Duration
+```
+
+```
+$ kubectl explain nodepool.spec.disruption.terminationGracePeriod
+KIND:     NodePool
+VERSION:  v1beta1
+
+FIELD:    terminationGracePeriod <metav1.Duration string>
+
+DESCRIPTION:
+     An optional maximum duration the node is provided to terminate gracefully. May be
+     decreased in delete request. Value must be a duration string ("30m", "24h", etc). A duration
+     of zero indicates that all pods should be terminated immediately, bypassing eviction
+     (including any PDBs). If this value is nil (default), deletion of the node will wait
+     indefinitely for pods to terminate gracefully before finalizing cleanup. Set this value
+     to ensure that nodes are deleted after a certain amount of time if your cluster operation
+     practices need it. Can also be used to provide a hard limit for users to extend the
+     run time of jobs during node deletion.
+```
+
+## Code Definition
+
+```go
+type Disruption struct {
+    {...}
+    // terminationGracePeriod is a nillable duration, parsed as an integer duration of number of seconds.
+    // A nil value means there is no timeout before pods will be forcibly evicted during node deletion.
+    terminationGracePeriod *NillableDuration `json:"terminationGracePeriod" hash:"ignore"`
```

Comment:
```suggestion
    terminationGracePeriod *metav1.Duration `json:"terminationGracePeriod" hash:"ignore"`
```

Seems like this could just be a standard `metav1.Duration` if we don't need the `Never` value

Reply:
As long as `metav1.Duration` can be nil we should be good to go.

Reply:
We use NillableDuration if there's a non-nil value that we want the field to default to.

Reply:
Updated to `metav1.Duration`.